Improved logging API

Lucas-C opened the pull request:
* [x] implementing then changes suggested in PR https://github.com/motis-project/utl/pull/25
* [x] switch from reStructuredText to Markdown in the `docs/`

## Documentation preview
![image](https://github.com/user-attachments/assets/64eb6cad-b095-41d4-b92d-a97d9cd9f983)

Lucas-C commented:
I have compiler issues like below, that I cannot figure how to solve properly...
```c++
error: class template argument deduction failed:
    9 |   log(llvl::info, "[{}] starting", name);
      |                                        ^
src/timer.cc:9:40: error: no matching function for call to ‘log(utl::llvl, const char [14], std::string&)’

include/utl/logging.h:51:5: note: candidate: ‘log(utl::llvl, fmt::v10::format_string<T ...>, const T&& ...)-> utl::log<T> [with T = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; fmt::v10::format_string<T ...> = fmt::v10::basic_format_string<char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >]’ (near match)
   51 |     log(llvl const level, fmt::format_string<T...> fmt_str, const T&&... args)
      |     ^~~
include/utl/logging.h:51:5: note:   conversion of argument 3 would be ill-formed:
src/timer.cc:9:36: error: cannot bind rvalue reference of type ‘const std::__cxx11::basic_string<char>&&’ to lvalue of type ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’}
    9 |   log(llvl::info, "[{}] starting", name);
      |                                    ^~~~
```

felixguendling commented:
Does this solve it?

https://gcc.godbolt.org/z/fxKTzYrve

```cpp
#include <source_location>
#include <iostream>
#include <iomanip>
#include <initializer_list>
#include <cinttypes>

#include "fmt/core.h"
#include "fmt/ostream.h"
#include "fmt/ranges.h"
#include "fmt/std.h"

namespace utl {

enum class log_level { debug, info, error };

log_level s_verbosity;

constexpr char const* to_str(log_level const level) {
  switch (level) {
    case log_level::debug: return "debug";
    case log_level::info: return "info";
    case log_level::error: return "error";
  }
  return "";
}

inline std::string now() {
  using clock = std::chrono::system_clock;
  auto const now = clock::to_time_t(clock::now());
  struct tm tmp {};
#if _MSC_VER >= 1400
  gmtime_s(&tmp, &now);
#else
  gmtime_r(&now, &tmp);
#endif

  std::stringstream ss;
  ss << std::put_time(&tmp, "%FT%TZ");
  return ss.str();
}


template <log_level LogLevel, typename... Args>
struct log {    
    log(char const* ctx, fmt::format_string<Args...> fmt_str, Args&&... args,
        std::source_location const& loc = std::source_location::current())
            : ctx_{ctx},
              args_{fmt_str, std::forward<Args>(args)...},
              loc_{loc}
    {}

    ~log() {
        if (LogLevel >= utl::s_verbosity) {
            fmt::print(std::clog, "{time} [{level}] [{file}:{line} {fn}] [{ctx}] {msg} {params}\n",
                    fmt::arg("file", loc_.file_name()),
                    fmt::arg("line", loc_.line()),
                    fmt::arg("fn", loc_.function_name()),
                    fmt::arg("time", now()),
                    fmt::arg("level", to_str(LogLevel)),
                    fmt::arg("ctx", ctx_),
                    fmt::arg("msg", std::apply([](fmt::format_string<Args...> fmt, Args... args) { return fmt::format(fmt, args...); }, args_)),
                    fmt::arg("params", params_));
        }
    }

    void params(std::initializer_list<std::pair<std::string_view, std::string_view>>&& params) {
        params_ = std::move(params);
    }
    
    char const* ctx_;
    std::tuple<fmt::format_string<Args...> , Args...> args_;
    std::source_location loc_;
    std::initializer_list<std::pair<std::string_view, std::string_view>> params_;
};

template <typename... Args>
struct info : public log<log_level::info, Args...> {
    using log<log_level::info, Args...>::log;
};

template <typename... Args>
info(char const*, fmt::format_string<Args...>, Args&&... args) -> info<Args...>;



template <typename... Args>
struct debug : public log<log_level::debug, Args...> {
    using log<log_level::debug, Args...>::log;
};

template <typename... Args>
debug(char const*, fmt::format_string<Args...>, Args&&... args) -> debug<Args...>;



template <typename... Args>
struct error : public log<log_level::error, Args...> {
    using log<log_level::error, Args...>::log;
};

template <typename... Args>
error(char const*, fmt::format_string<Args...>, Args&&... args) -> error<Args...>;

}  // namespace utl

int main() {
  auto dataset = "test";
  utl::info("nigiri.loader", "starting import: dataset={}", dataset).params({{"dataset", dataset}});
  utl::error("nigiri.loader", "error import: dataset={}", dataset).params({{"dataset", dataset}, {"file", "stops.txt"}});
  utl::debug("nigiri.loader", "debug message");
}
```

Lucas-C commented:
I figured it out!
The GitHub Actions pipeline should be passing now.
Thank you very much for the pointers you gave me @felixguendling 👍

For now, I did not include the `loc_.function_name()` in the log line,
because it can be quite long and IMHO it makes logs harder to read
without providing much valuable information.

A few API points remain to be discussed and decided upon:

### Mandatory `ctx` VS optional `.ctx()` call
I do not have any strong opinion about this: what do you think is the best?

### `llvl` VS `log_level`
For something that will be repeated very often, I have a tendency to prefer the shorter `llvl`.

### `.metadata()` VS `.params()`
And should they be inserted in the final full log line?

I do not quite understand the point of having `params` when "dynamic" variables can already be passed as optional parameters.

In my original idea, metadata provided would only be used when producing structured logs, that could possibly be sent to tools like DataDog.

What usage do you have in mind for `params`?

Lucas-C commented:
Damn 🤦

Though all works fine with MSVC,
wiyth GCC & Clang I still have a compilation error for this template deduction guide:
```c++
template <llvl LogLevel, typename... Args>
log(fmt::format_string<Args...>, Args&&... args) -> log<LogLevel, Args...>;
```

```c++
include/utl/logging.h:115:1: error: deduction guide template contains a template parameter that cannot be deduced
  115 | log(fmt::format_string<Args...>, Args&&... args) -> log<LogLevel, Args...>;
      | ^
include/utl/logging.h:114:16: note: non-deducible template parameter 'LogLevel'
  114 | template <llvl LogLevel, typename... Args>
      |                ^
```

felixguendling commented:
> Damn 🤦
> 
> Though all works fine with MSVC, wiyth GCC & Clang I still have a compilation error for this template deduction guide:
> 
> ```c++
> template <llvl LogLevel, typename... Args>
> log(fmt::format_string<Args...>, Args&&... args) -> log<LogLevel, Args...>;
> ```
> 
>  ```c++
> include/utl/logging.h:115:1: error: deduction guide template contains a template parameter that cannot be deduced
>   115 | log(fmt::format_string<Args...>, Args&&... args) -> log<LogLevel, Args...>;
>       | ^
> include/utl/logging.h:114:16: note: non-deducible template parameter 'LogLevel'
>   114 | template <llvl LogLevel, typename... Args>
>       |                ^
> ```

The problem is that the `LogLevel` cannot be inferred from the constructor usage itself. I think you can just remove the deduction guide for `log` itself. The deduction guides for `error`, `info`, and `debug` should be sufficient.

felixguendling commented:
> For now, I did not include the `loc_.function_name()` in the log line, because it can be quite long and IMHO it makes logs harder to read without providing much valuable information.

It helps to know the cause without checking which function is in the given line in the given file. We can put it as a default parameter.



> A few API points remain to be discussed and decided upon:
> 
> ### Mandatory `ctx` VS optional `.ctx()` call
> I do not have any strong opinion about this: what do you think is the best?

I know for sure that if it's not mandatory, it will rarely be used. That's probably fine if we use OTEL spans instead. In this case, the information would be redundant anyway.


> ### `llvl` VS `log_level`
> For something that will be repeated very often, I have a tendency to prefer the shorter `llvl`.

As far as I understand the API now, `log_level` would never be used in client code because only `utl::info`, `utl::error`, and `utl::debug` (maybe we add `utl::warn`?) will be used directly. I consider `log_level` an implementation detail which can be written with its full name.

> ### `.metadata()` VS `.params()`
> And should they be inserted in the final full log line?

I am new to OTEL but maybe we stick to their terminology. If I understand OTEL correctly, they call them `attributes`. Maybe we call the function `.attr()`? 

> I do not quite understand the point of having `params` when "dynamic" variables can already be passed as optional parameters.
> 
> In my original idea, metadata provided would only be used when producing structured logs, that could possibly be sent to tools like DataDog.
> 
> What usage do you have in mind for `params`?

My idea was that it's easy to use `grep` and other command line tools to quickly find everything connected to a dataset, a client IP, etc. - this is easier if the parameters are part of the log line and are given in a consistent format. This is not the case for the message the user assembles (which is not really made to be machine readable - it's for humans).

Lucas-C commented:
> I think you can just remove the deduction guide for log itself. The deduction guides for error, info, and debug should be sufficient.

That worked, but there is one remaining issue with the MacOSX build:
https://github.com/motis-project/utl/actions/runs/12816832215/job/35738669646?pr=27
It seems that `std::source_location::current()` returns `logging.h` on this platform,
and not the calling file path.

> It helps to know the cause without checking which function is in the given line in the given file. We can put it as a default parameter.

Alright.

> I know for sure that if it's not mandatory, it will rarely be used. That's probably fine if we use OTEL spans instead. In this case, the information would be redundant anyway.

Alright, I made it mandatory.

> I consider log_level an implementation detail which can be written with its full name.

Alright.

> I am new to OTEL but maybe we stick to their terminology. If I understand OTEL correctly, they call them attributes. Maybe we call the function .attr()?

Let's go for it.

MichaelKutzner commented:
I just tried to find out why the macOS build failed. Apparently the error is caused by a bug in Clang 15 (https://github.com/llvm/llvm-project/issues/56379), which has been fixed with Clang 16 and couldn't be backported easily. So it still affects macOS build, which seems to be based on Clang 15.
Maybe these test should be disabled, when compiled with Clang 15?

Lucas-C commented:
> I just tried to find out why the macOS build failed. Apparently the error is caused by a bug in Clang 15 ([llvm/llvm-project#56379](https://github.com/llvm/llvm-project/issues/56379)), which has been fixed with Clang 16 and couldn't be backported easily. So it still affects macOS build, which seems to be based on Clang 15. Maybe these test should be disabled, when compiled with Clang 15?

Thank you for digging & finding this explanation! 👍

I modified the tests to not check the filename,
and added a comment in the source code about that.

Lucas-C commented:
I think this PR is ready to be reviewed & merged.

Lucas-C commented:
I opened issue https://github.com/motis-project/utl/issues/28 to track further improvements that could be made in later PRs.

felixguendling commented on include/utl/logging.h:
```diff
@@ -57,46 +45,83 @@ inline std::string now() {
+
+  ~log() {
+    if (LogLevel >= utl::s_verbosity) {
+#if defined(_WIN32)
+      const char* baseFileName = strrchr(loc_.file_name(), '\\')
+                                     ? strrchr(loc_.file_name(), '\\') + 1
+                                     : loc_.file_name();
+#else
+      // On MacOS, due to a bug with Clang 15, the wrong filename
+      // is retrieved (logging.h instead of the calling file):
+      // https://github.com/llvm/llvm-project/issues/56379
+      const char* baseFileName = strrchr(loc_.file_name(), '/')
+                                     ? strrchr(loc_.file_name(), '/') + 1
+                                     : loc_.file_name();
+#endif
```
We use `snake_case` for local variables.

We try to make everything `auto`. This makes it obvious when you do implicit casts (because you would have to write `static_cast` on the right side of the assignment).
https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

felixguendling replied:
Oops. `auto const` to be precise.

Lucas-C replied:
Alright, done! 🙂

felixguendling commented on docs/index.md:
```diff
@@ -0,0 +1,49 @@
+# MOTIS utl module documentation
```
Maybe mention that those logging function are an exception to the rule that we use [Aggregate Initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization) wherever possible. In principle you could write `utl::info{...}` which is not what we want.

Lucas-C replied:
I added a note:
![image](https://github.com/user-attachments/assets/bd214902-cbdd-4837-86ac-2291edd36fd6)

felixguendling commented on include/utl/logging.h:
```diff
@@ -57,46 +45,83 @@ inline std::string now() {
 
-}  // namespace utl
+  void attrs(
+      std::initializer_list<std::pair<std::string_view, std::string_view> >&&
```
Maybe add to the "future work" issue that we could support more datatypes by making it a `std::pair<std::string_view, std::variant</* ... all supported value types */>>`. This would require a custom formatter for libfmt but it's doable.

Lucas-C replied:
Alright, added.

felixguendling commented:
🥳
